File: programmers/모두0으로만들기.cpp
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   모두0으로만들기.cpp                                    :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: jaeelee <jaeelee@student.42seoul.kr>       +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2021/10/03 22:24:01 by jaeelee           #+#    #+#             */
/*   Updated: 2021/10/03 22:24:23 by jaeelee          ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

/*
   프로그래머스 4월 코드챌린지 문제

   트리의 모든 점의 가중치를 0으로 만들기
   임의의 연결된 두 점을 골라서 한쪽은 1증가시키고, 다른 한쪽은 1 감소시킨다.
 */

#include <bits/stdc++.h> // 자수 사용하는 라이브러리들을 컴파일하도록 함
using namespace std;

long long answer = 0;
vector< vector<int> > maps;
vector<long long> weight;
vector<int> visited;

void dfs(int x, int top)
{
	for (int i : maps[x])
	{
		if (i == top) continue; // 위치 중요!
		if (maps[i].size() != 1) dfs(i, x); // 위치 중요!
		answer += abs(weight[i]);
		weight[x] += weight[i];
		weight[i] = 0;
	}
}

long long solution(vector<int> a, vector< vector<int> > edges) {
	int i;
	int sum;

	visited.assign(a.size(), 0);
	maps.assign(a.size(), vector<int> (0));
	for (int i : a) weight.push_back((long long)i);
	sum = 0;
	i = -1;
	while (++i < a.size())
		sum += a[i];
	if (sum != 0)
		return (-1);

	i = -1;
	while (++i < edges.size())
	{
		maps[edges[i][0]].push_back(edges[i][1]);
		maps[edges[i][1]].push_back(edges[i][0]);
	}

	dfs(0, -1);
	return answer;
}

File: programmers/모두0으로만들기.py
# **************************************************************************** #
#                                                                              #
#                                                         :::      ::::::::    #
#    모두0으로만들기.py                                     :+:      :+:    :+:    #
#                                                     +:+ +:+         +:+      #
#    By: jaeelee <jaeelee@student.42seoul.kr>       +#+  +:+       +#+         #
#                                                 +#+#+#+#+#+   +#+            #
#    Created: 2021/09/24 18:44:55 by jaeelee           #+#    #+#              #
#    Updated: 2021/09/24 18:44:57 by jaeelee          ###   ########.fr        #
#                                                                              #
# **************************************************************************** #

"""
	프로그래머스 4월 코드챌린지 문제

    트리의 모든 점의 가중치를 0으로 만들기
    임의의 연결된 두 점을 골라서 한쪽은 1증가시키고, 다른 한쪽은 1 감소시킨다.
"""


import sys
sys.setrecursionlimit(300000) # 재귀 깊이 제한 참고 : https://fuzzysound.github.io/sys-setrecursionlimit


def dfs(a, graph, start, visited):
	global answer
	visited[start] = 1
	ret = 0
	for i in graph[start]:
	if visited[i] == 1:
	continue
	if len(graph[i]) != 1: # 리프노드가 아닐 경우
	dfs(a, graph, i, visited)
	answer += abs(a[i])
	a[start] += a[i]
	a[i] = 0


	def solution(a, edges):
		global answer
		answer = 0
		visited = [0 for i in range(len(a))]
		graph = [[] for i in range(len(a))]
		if sum(a) != 0:
		return -1

		for i in range(len(edges)):
	graph[edges[i][0]].append(edges[i][1])
	graph[edges[i][1]].append(edges[i][0])

	dfs(a, graph, 0, visited)
	return answer
